game-core/src/main/java/games/strategy/triplea/delegate/battle/FireAa.java

```diff
@@ -92,84 +89,77 @@
   public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
 
     // Loop through each type of AA and break into firing groups based on suicideOnHit
-    for (final String aaType : aaTypes) {
-      final Collection<Unit> aaTypeUnits =
-          CollectionUtils.getMatches(firingUnits, Matches.unitIsAaOfTypeAa(aaType));
-      final List<Collection<Unit>> firingGroups = MustFightBattle.newFiringUnitGroups(aaTypeUnits);
-      for (final Collection<Unit> firingGroup : firingGroups) {
-        final Set<UnitType> validTargetTypes =
-            UnitAttachment.get(firingGroup.iterator().next().getType())
-                .getTargetsAa(bridge.getData());
-        final Set<UnitType> airborneTypesTargeted =
-            defending
-                ? TechAbilityAttachment.getAirborneTargettedByAa(hitPlayer, bridge.getData())
-                    .get(aaType)
-                : new HashSet<>();
-        final Collection<Unit> validTargets =
-            CollectionUtils.getMatches(
-                attackableUnits,
-                Matches.unitIsOfTypes(validTargetTypes)
-                    .or(
-                        Matches.unitIsAirborne()
-                            .and(Matches.unitIsOfTypes(airborneTypesTargeted))));
-        final IExecutable rollDice =
-            new IExecutable() {
-              private static final long serialVersionUID = 6435935558879109347L;
+    final List<FiringGroup> groupsAndTargets =
+        Aa.builder()
+            .aaUnits(firingUnits)
+            .aaTypes(aaTypes)
+            .hitPlayer(hitPlayer)
+            .attackableUnits(attackableUnits)
+            .defending(defending)
+            .gameData(bridge.getData())
+            .build()
+            .getFiringGroups();
 
-              @Override
-              public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
-                validTargets.removeAll(casualtiesSoFar);
-                if (!validTargets.isEmpty()) {
-                  dice =
-                      DiceRoll.rollAa(
-                          validTargets,
-                          firingGroup,
-                          allEnemyUnitsAliveOrWaitingToDie,
-                          allFriendlyUnitsAliveOrWaitingToDie,
-                          bridge,
-                          battleSite,
-                          defending);
-                  if (!headless) {
-                    SoundUtils.playFireBattleAa(firingPlayer, aaType, dice.getHits() > 0, bridge);
-                  }
+    for (final FiringGroup groupAndTarget : groupsAndTargets) {
+      final Collection<Unit> validTargets = groupAndTarget.getValidTargets();
+      final Collection<Unit> firingGroup = groupAndTarget.getFiringGroup();
```

Comment:
It's a bit confusing to have a FiringGroup have a member called firingGroup and a getter for it.

Could it be named differently? e.g. getFiringUnits()?